データベース設計

ganaha-mikoto opened the pull request:
# What
ER図を作成
READMEにデータベース設計を記述

# Why
フリマアプリの機能を考え、必要なテーブル・カラム情報を洗い出し、データベースを作成するための設計をするため

ER図のGyazo↓↓↓
https://gyazo.com/62768ac21cf76546f0a3b3ff90bd72cd

ganaha-mikoto committed:
データベース設計

フリマアプリの機能を考え、必要なテーブル・カラム情報を洗い出す。
ER図を書く。
READMEにデータベース設計を記述する。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,56 @@
+| condition           | string     | null: false, default: '---'    |
+| shipping_charge     | string     | null: false, default: '---'    |
+| shipping_area       | string     | null: false, default: '---'    |
+| shipping_days       | string     | null: false, default: '---'    |
```
カテゴリー・商品の状態・配送料負担・発送元地域・発送までの日数カラムはActivehashを用いた実装になるため、Activehashのルールに沿ったカラム名と型にしましょう。
activehashにつきましては以下カリキュラムをご参考にいただければと思います。

```
Lesson1 最終課題補足カリキュラム
カテゴリーの選択を工夫しよう
```

yuri240 commented on README.md:
```diff
@@ -1,24 +1,56 @@
+  - has_many :products
+  - has_many :purchasers
+
+## products テーブル
```
ユーザーの外部キーカラムを追加しましょう。

■理由
実装要件において、商品テーブルには出品者が誰であるかを保存することが求められているためです。

yuri240 commented on README.md:
```diff
@@ -1,24 +1,56 @@
+| building_name       | string     |                                |
+| phone_number        | string     | null: false                    |
+| user_id             | integer    | null: false, foreign_key: true |
+| product_id          | integer    | null: false, foreign_key: true |
```
現状の設計ですと、purchasersテーブルに`配送先住所`と`購入履歴`を保存する設計となっているようです。

今回のフリマアプリにおいては、
以下2つのテーブルに分けて設計していただけると良さそうです。

```
①配送先住所を登録するテーブル
②購入履歴を登録するテーブル
```

また、`②購入履歴を登録するテーブル`には、「誰が」「どの商品を購入したか」を保存するためにユーザーと商品の外部キーカラムのみを用意しましょう。

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
初回から完成度高く設計できており、素晴らしいですね。

コメントさせていただきましたので、ご確認お願いいたします。

また、次回ご提出の際にはER図も修正が必要となります。

お手数ではありますが、
修正後のER図を再度こちらのプルリクエストに添付していただくようお願いいたします。

ganaha-mikoto committed:
テーブル追加と名称変更

・外部キーの追加
・Activehashのルールに沿ったカラム名と型に変更
・配送先住所と購入履歴を別々のテーブルにした

ganaha-mikoto commented:
# What
テーブルの追加と名称変更

# why
・外部キーを追加し出品者と商品を各テーブルと紐付けするため
・Activehashのルールに沿ったカラム名と型に変更し、選択式のデータにするため
・配送先住所と購入履歴を別々のテーブルにするため

## ER図
[https://gyazo.com/6b6ed78dfef578b924e155892c07a895](url)

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,74 @@
-* Ruby version
+### Association
+  - has_many :products
+  - has_many :purchasers
```
```ruby
  - has_many :purchasers
 ```
users テーブルの上記のアソシエーションは不要かと存じます。
削除しましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,74 @@
+| condition_id        | integer    | null: false, default: 0        |
+| shipping_charge_id  | integer    | null: false, default: 0        |
+| shipping_area_id    | integer    | null: false, default: 0        |
+| shipping_days_id    | integer    | null: false, default: 0        |
```
```ruby
| shipping_days_id    | integer    | null: false, default: 0        |
```
active_hashのカラム名は`(単数形)_id`としておきましょう。

モデルを作成する際にモデル名は単数形の原則に反してしまうためです。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,74 @@
+| shipping_area_id    | integer    | null: false, default: 0        |
+| shipping_days_id    | integer    | null: false, default: 0        |
+| price               | integer    | null: false                    |
+| user_id             | integer    | null: false, foreign_key: true |
```
```ruby
| user_id             | integer    | null: false, foreign_key: true |
```
外部キーはまずデータ型は`references`としておきましょう。
またカラム名の`_id`は自動で付与されるため削除しましょう。

他の外部キーも同様の修正をお願いいたします。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,74 @@
-* How to run the test suite
+  - belongs_to :users
+  - has_one :purchasers
+  - has_one :histories
```
```ruby
  - belongs_to :users
  - has_one :purchasers
  - has_one :histories
 ```
上記のアソシエーションに関して
`  - has_one :purchasers`は削除して良いかと存じます。
`belongs_to`や`has_one`の後は単数形にしましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,74 @@
+| address             | string     | null: false                    |
+| building_name       | string     |                                |
+| phone_number        | string     | null: false                    |
+| user_id             | integer    | null: false, foreign_key: true |
+| product_id          | integer    | null: false, foreign_key: true |
```
```ruby
| user_id             | integer    | null: false, foreign_key: true |
| product_id          | integer    | null: false, foreign_key: true |
```
上記の外部キーを削除し、`histories テーブル`に関する外部キーを追加しましょう。

histories テーブルとaddresses テーブルを紐づけておけば「誰が」「何を」購入し「どこに」届けるのか、の情報を全て取得ができるためです。

この修正に伴い、histories テーブルとaddresses テーブルにアソシエーションの修正をお願いいたします。

ganaha-mikoto replied:
申し訳ないですが、この修正指示の認識の確認したいです。

addressessテーブルとhistoriesテーブルを紐付け、重複してuser_idとproduct_idを呼び出さないように、historiesテーブルからuser_idとproduct_idを呼び出せるようにするという認識ですがあっていますか？

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます。

コメントいたしましたので、ご確認をお願いいたします。

コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りください。
ご対応のほど、よろしくお願いいたします。

ganaha-mikoto committed:
カラムの変更とhistoriesテーブルの紐付け

usersテーブルのpurchasers  テーブルのアソシエーションの削除
active_hashのカラム名を単数形に変更
belongs_toやhas_oneの後を単数形に変更
histories テーブルとaddresses テーブルを紐づけとhistories テーブルとaddresses テーブルにアソシエーションの修正

ganaha-mikoto reviewed (commented):
# what
usersテーブルのpurchasers  テーブルのアソシエーションの削除
active_hashのカラム名を単数形に変更
belongs_toやhas_oneの後を単数形に変更
histories テーブルとaddresses テーブルを紐づけとhistories テーブルとaddresses テーブルにアソシエーションの修正

# Why
不要なテーブルやカラムを削除して無駄な機能をなくす
historyテーブルを利用した各テーブルの関連付けを効率的にするため

## ER図
[https://gyazo.com/c04316a0f354d51eea164a30497bdc32](url)

wad-sho commented on README.md:
```diff
@@ -1,24 +1,75 @@
+| ------------------- | ---------- | ------------------------------ |
+| user                | references | null: false, foreign_key: true |
+| product             | references | null: false, foreign_key: true |
+| address             | references | null: false, foreign_key: true |
```
> addressessテーブルとhistoriesテーブルを紐付け、重複してuser_idとproduct_idを呼び出さないように、historiesテーブルからuser_idとproduct_idを呼び出せるようにするという認識ですがあっていますか？

上記ご認識の通りになります。

historiesテーブルにとってaddressesテーブルは子になりますので、以下は削除しましょう！
`| address             | references | null: false, foreign_key: true |`

wad-sho commented on README.md:
```diff
@@ -1,24 +1,75 @@
+| Column              | Type       | Options                        |
+| ------------------- | ---------- | ------------------------------ |
+| postal_code         | string     | null: false                    |
+| prefecture          | string     | null: false                    |
```
| prefecture          | string     | null: false                    |

こちらのカラムは以下に修正しましょう！

| shipping_area_id    | integer    | null: false, default: 0        |

Activehashは同じデータを別テーブルにも兼用できるためです。

wad-sho commented on README.md:
```diff
@@ -1,24 +1,75 @@
-* Ruby version
+### Association
+  - has_many :products
+  - has_many :purchasers
```
has_many :purchasers

上記不要となりますので削除しておきましょう。

wad-sho commented on README.md:
```diff
@@ -1,24 +1,75 @@
+
+  - belongs_to :user
+  - belongs_to :product
+  - has_one :history
```
```
  - belongs_to :user
  - belongs_to :product
  - has_one :history
```

has_one :historyのみが必要なアソシエーションになります。
修正しましょう。

ganaha-mikoto replied:
これはhistoriesテーブルでuser_idとproduct_idを保存しており、history_idを関連付けているためアソシエーションがなくても呼び出せるということですか

Akinori0123 replied:
◎addressesテーブルのアソシエーション
```
- has_one :history
```
前回レビュー時に一点見落としており失礼しました。
addressesテーブルにはhistoryの外部キーカラムがありますので、`has_one: history`を`belongs_to :history`に変更しましょう。


> これはhistoriesテーブルでuser_idとproduct_idを保存しており、history_idを関連付けているためアソシエーションがなくても呼び出せるということですか

ご記載いただいた認識でバッチリです。
「誰が」「どの商品を購入したか」はhistoriesテーブルで管理されているため、addressesテーブルにuserやproductとのアソシエーションは不要になります。

wad-sho reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。

コメントいたしましたので、ご確認頂けますと幸いでございます。
もしレビューのコメントや内容にご不明点等ある場合はご遠慮なくテックコネクトをご活用ください。
（レビュアーは質問対応ができかねてしまうためです。申し訳ございません）

指摘内容の修正がお済みになりましたら、再度レビュー依頼くださいませ。
その際、ER図もREADMEの内容と合わせて修正とお願いいたします。
**次回のレビュー依頼の際に最新のER図のURLをまたこちらのプルリクエストに添付頂お願いいたします。**

ganaha-mikoto committed:
アソシエーションの削除

historyテーブルと呼び出しが重なっているアソシエーションを修正
active hashを他のテーブルでも利用できるように同じ項目は統一した

ganaha-mikoto reviewed (commented):
# What
Activehashのカラム修正
historyテーブルに関連するアソシエーションの修正

# Why
Activehashgは同じデータを他のテーブルでも扱えるため
テーブルをアソシエーションで繋がなくてもhistoriesテーブルが中間テーブルの代わりとなるため

## ER図
[https://gyazo.com/105c0dedcee9dcc3c74983476d3730be](url)

TakashiKozuma commented:
レビュー依頼ありがとうございます！

いただいたレビュー依頼ですが、修正点の差分が確認できない状況です。
コミットとプッシュのし忘れがないかをご確認いただき、修正点の差分を出して再度ご提出をお願いいたします。

ganaha-mikoto commented:
[https://github.com/ganaha-mikoto/furima-40447/pull/14/commits/db0d0086c0f498f7a4abb22b2f24b854542fbcb9](url)

自身でも確認しましたが差分は出ていますし、commitもpublicできていました。何が悪かったのか自分ではわからないので教えて下さい。

ganaha-mikoto commented:
メンターに問題教えてもらいました。
お騒がせしました。
これからはWhatとWhyは最初だけ投稿します

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
前回の修正内容確認いたしました。追加でコメントいたしましたのでご確認のほどよろしくお願いいたします。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| condition_id        | integer    | null: false, default: 0        |
+| shipping_charge_id  | integer    | null: false, default: 0        |
+| shipping_area_id    | integer    | null: false, default: 0        |
+| shipping_day_id     | integer    | null: false, default: 0        |
```
```
default: 0 
```
追加で失礼いたします。
default: 0ですと空であった場合にデフォルトで0が保存されるようになりますが、activehashのデータにおいて0は異常値になります。
そのためdefault: 0のオプションは削除しましょう。

Akinori0123 commented on README.md:
```diff
@@ -1,24 +1,71 @@
+| Column              | Type       | Options                        |
+| ------------------- | ---------- | ------------------------------ |
+| postal_code         | string     | null: false                    |
+| shipping_area_id    | integer    | null: false, default: 0        |
```
```
| shipping_area_id    | integer    | null: false, default: 0        |
```
上記もdefault: 0のオプションを削除しましょう。

Akinori0123 reviewed (commented):
先ほどはレビュー依頼いただきありがとうございました！
分けてのレビューになってしまい申し訳ございません、追加の指摘内容がございましたので再度レビューを行わせていただきます。
併せてご確認のほどよろしくお願いいたします。

ganaha-mikoto committed:
不要なオプションの削除

activehashのdefault: 0のオプション削除

minachanchan commented on README.md:
```diff
@@ -1,24 +1,71 @@
 
-* How to run the test suite
+  - belongs_to :user
+  - belongs_to :history
```
```ruby
belongs_to :history
```
一つの購入履歴を持つというアソシエーションに変更しましょう！

minachanchan commented on README.md:
```diff
@@ -1,24 +1,71 @@
+
+### Association
+
+  - has_one :history
```
```ruby
has_one :history
```
[こちら](https://github.com/ganaha-mikoto/furima-40447/pull/14#discussion_r1648797370)改めてご確認いただき、`belongs_to :history`に変更しましょう。

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
着実に完成に近づいておりますね！

コメントいたしましたので、ご確認頂けますと幸いでございます。

指摘内容の修正がお済みになりましたら、再度レビュー依頼くださいませ。
その際、ER図もREADMEの内容と合わせて修正とお願いいたします。
**次回のレビュー依頼の際に最新のER図のURLをまたこちらのプルリクエストに添付頂ければと存じます。**

実装完了まであと少し、一緒に頑張りましょう！
何卒よろしくお願いいたします。

ganaha-mikoto committed:
addressテーブルのアソシエーション

ganaha-mikoto reviewed (commented):
ER図です
[https://gyazo.com/8ebc4ae98df97dc640af384f52b9a4c2](url)

TakashiKozuma commented:
追加のレビュー依頼ありがとうございます！
修正点も含め、確認が完了しました。
LGTMです！！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に次の実装のためのブランチを切ってから次にお進み下さい。
